zanata-war/src/main/java/org/zanata/liquibase/custom/ChangePositionalResIdToContentHash.java

```diff
@@ -0,0 +1,224 @@
+import org.zanata.common.DocumentType;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.HashSet;
+import java.util.Set;
+
+
+/**
+ * Change the resource ID (resId) for text flows in some document formats to be
+ * a hash of the content, rather than a position-based value.
+ *
+ * Text flows in the following document types {@link org.zanata.common.DocumentType} are
+ * modified:
+ *
+ *   - txt {@link org.zanata.common.DocumentType#PLAIN_TEXT}
+ *   - all libreoffice types
+ *
+ * If multiple text flows in the same document have the same content, only the first
+ * will be kept.
+ */
+public class ChangePositionalResIdToContentHash implements CustomTaskChange {
+
+    private static final Set<DocumentType> documentTypes = ImmutableSet.of(DocumentType.PLAIN_TEXT,
+            DocumentType.OPEN_DOCUMENT_DATABASE, DocumentType.OPEN_DOCUMENT_FORMULA,
+            DocumentType.OPEN_DOCUMENT_GRAPHICS, DocumentType.OPEN_DOCUMENT_PRESENTATION,
+            DocumentType.OPEN_DOCUMENT_SPREADSHEET, DocumentType.OPEN_DOCUMENT_TEXT);
+
+    private static final String countDocumentsOfTypeSql =
+            "select count(*) from HDocument_RawDocument " +
+                    "left join HRawDocument on HDocument_RawDocument.rawDocumentId=HRawDocument.id " +
+                    "where HRawDocument.type = ?";
+
+    private static final String selectDocumentsOfTypeSql =
+            "select documentId from HDocument_RawDocument " +
+                    "left join HRawDocument on HDocument_RawDocument.rawDocumentId=HRawDocument.id " +
+                    "where HRawDocument.type = ?";
+
+    // id must be present as primary key to allow updates in the ResultSet
+    private static final String selectTextFlowsSql =
+            "select id, contentHash, pos, resId, obsolete from HTextFlow where document_id = ? " +
+                    "order by pos";
+    private static final int ID_COLUMN = 1;
+    private static final int CONTENT_HASH_COLUMN = 2;
+    private static final int POS_COLUMN = 3;
+    private static final int RES_ID_COLUMN = 4;
+    private static final int OBSOLETE_COLUMN = 5;
+
+    private Logger log = LogFactory.getLogger();
+
+    private PreparedStatement countDocumentsOfType;
+    private PreparedStatement selectDocumentsOfType;
+    private PreparedStatement selectTextFlows;
+
+    private int totalDocs = 0;
+    private int totalTextFlows = 0;
+
+    @Override
+    public void execute(Database database) throws CustomChangeException {
+
+        try {
+            prepareStatements((JdbcConnection) database.getConnection());
+            // could use try-with-resources to define the PreparedStatements here
+            // but making them fields allows for cleaner code.
+            try {
+                for (DocumentType type : documentTypes) {
+                    migrateResId(type);
+                }
+            } finally {
+                closeStatements();
```

Comment:
This won't be called if there is an exception preparing the second or third statements.  I think you should use try with resources, but you can still leave them as fields.